### ISSUE TITLE
Overloading `__init__` doesn't behave the same as overloading `__new__`

### ISSUE THREAD

**Issue by sbdchd** [microsoft/pyright#1550]
**Describe the bug**

Overloading `__init__`'s `self` param with a generic argument doesn't work as expected, but
if I instead `overload` `__new__` the types work.

**To Reproduce**

When I `overload` the `__init__` method with the `self` param typed, the `null=True` `overload`
doesn't result in the type being `T | None`, instead it's `T`:

```python
from __future__ import annotations

from typing import Any, Generic, Optional, Type, TypeVar, overload

from typing_extensions import Literal


class Model:
    ...


_T = TypeVar("_T", bound="Optional[Model]")


class ForeignKey(Generic[_T]):
    @overload
    def __init__(
        self: ForeignKey[_T], to: Type[_T], *, null: Literal[False] = ...
    ) -> None:
        ...

    @overload
    def __init__(
        self: ForeignKey[Optional[_T]], to: Type[_T], *, null: Literal[True]
    ) -> None:
        ...

    def __init__(self, to: Type[_T], *, null: bool = False) -> None:
        raise NotImplementedError

    @overload
    def __get__(self: ForeignKey[_T], instance: Any, owner: Any) -> _T:
        ...

    @overload
    def __get__(
        self: ForeignKey[Optional[_T]], instance: Any, owner: Any
    ) -> Optional[_T]:
        ...

    def __get__(self, instance: object, owner: object) -> Any:
        raise NotImplementedError


class Author(Model):
    pass


class Post(Model):
    author = ForeignKey(Author, null=False)
    co_author = ForeignKey(Author, null=True)


post = Post()
reveal_type(post.author)
# pylance: Type of "Post.author" is "Author"
# mypy output: Revealed type is 'Author'
reveal_type(post.co_author)
# pylance: Type of "post.co_author" is "Author"
# mypy output: Revealed type is 'Union[Author, None]'
```

However, if I instead `overload` the `__new__` method, the types work as expected:

```python
from __future__ import annotations

from typing import Any, Generic, Optional, Type, TypeVar, overload

from typing_extensions import Literal


class Model:
    ...


_T = TypeVar("_T", bound="Optional[Model]")


class ForeignKey(Generic[_T]):
    @overload
    def __new__(cls, to: Type[_T], *, null: Literal[False] = ...) -> ForeignKey[_T]:
        ...

    @overload
    def __new__(cls, to: Type[_T], *, null: Literal[True]) -> ForeignKey[Optional[_T]]:
        ...

    def __new__(cls, to: Type[_T], *, null: bool = False) -> ForeignKey[Any]:
        raise NotImplementedError

    @overload
    def __get__(self: ForeignKey[_T], instance: Any, owner: Any) -> _T:
        ...

    @overload
    def __get__(
        self: ForeignKey[Optional[_T]], instance: Any, owner: Any
    ) -> Optional[_T]:
        ...

    def __get__(self, instance: object, owner: object) -> Any:
        raise NotImplementedError


class Author(Model):
    pass


class Post(Model):
    author = ForeignKey(Author, null=False)
    co_author = ForeignKey(Author, null=True)


post = Post()

reveal_type(post.author)
# pylance: Type of "Post.author" is "Author"
# mypy output: Revealed type is 'Author'
reveal_type(post.co_author)
# pylance: Type of "Post.co_author" is "Author | None"
# mypy output: Revealed type is 'Union[Author, None]'
```

pyright config:

```json
{
  "typeCheckingMode": "strict",
  "reportMissingModuleSource": false
}
```

**Expected behavior**

The `null=True` `overload` results in the same behavior as overloading via `__new__`

**Screenshots or Code**
see above code

**VS Code extension or command-line**
pyright 1.1.74

**Additional context**
Issue seems similar to: https://github.com/microsoft/pyright/issues/1211

Also pylance 2021.2.4 reports a lot more false positives, but I'm assuming it's vendoring an older version of pyright.

**Comment by erictraut** [comment]
The mechanism by which you can annotate the `self` parameter in an `__init__` method is a bit of a hack to begin with. It's not contemplated by PEP 484 or documented in any other PEPs. It's a convention that was decided upon by mypy, and it's (IMO) a pretty ugly solution. Normally, the "solved" TypeVars in a generic function are applied to the return type. But in the case of `__init__`, the return type is always `None`, hence the need for this odd exception.

I was reluctant to implement it in pyright unless/until it was fully spec'ed and ratified as part of the standard. I eventually implemented it based on a mini spec that I constructed. I shared that spec with the typeshed and mypy folks, and they seemed to think it was OK. In that spec, I indicated that pylance would support only the case that the `self` annotation was a concrete version of the class being constructed. You are using a generic version of the class in your example, which explains why it doesn't work in pyright today.

I've extended the logic to support generic versions, but there's one important caveat. If you use a TypeVar in the `self` annotation, it must be scoped to the function, not to the class. In your example above, you're using the annotation `ForeignKey[Optional[_T]]`, but `_T` is a class-scoped TypeVar. That presents a contradiction because the class is of type `ForeignKey[_T]`, but you are annotating the `self` parameter in an instance method to be `ForeignKey[Optional[_T]]`, which will never be compatible with the class. The correct workaround is to use a function-scoped TypeVar, like this:

```python
_T1 = TypeVar("_T1", bound="Optional[Model]")
_T2 = TypeVar("_T2", bound="Optional[Model]")

class ForeignKey(Generic[_T1]):
    @overload
    def __init__(
        self: "ForeignKey[_T2]", to: Type[_T2], *, null: Literal[False] = ...
    ) -> None:
        ...

    @overload
    def __init__(
        self: "ForeignKey[Optional[_T2]]", to: Type[_T2], *, null: Literal[True]
    ) -> None:
        ...
```

With my changes to pyright and the above changes to your code, this should now work.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.116, which I just published. It will also be in the next release of pylance.
